Source/WTF/wtf/NativePromise.h

```diff
@@ -885,49 +880,41 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         using DispatcherRealType = typename RemoveSmartPointer<DispatcherType>::type;
         static_assert(LooksLikeRCSerialDispatcher<DispatcherRealType>::value, "Must be used with a RefCounted SerialFunctionDispatcher");
 
-        using R1 = typename RemoveSmartPointer<typename MethodTrait<ResolveFunction>::returnType>::type;
-        using R2 = typename RemoveSmartPointer<typename MethodTrait<RejectFunction>::returnType>::type;
+        using R1 = typename RemoveSmartPointer<decltype(invokeWithVoidOrWithArg(std::forward<ResolveFunction>(resolveFunction), std::declval<std::conditional_t<std::is_void_v<ResolveValueT>, std::nullptr_t, ResolveValueT>>()))>::type;
+        using R2 = typename RemoveSmartPointer<decltype(invokeWithVoidOrWithArg(std::forward<RejectFunction>(rejectFunction), std::declval<RejectValueT>()))>::type;
         using IsChaining = std::bool_constant<RelatedNativePromise<R1, R2>>;
         static_assert(IsChaining::value || (std::is_void_v<R1> && std::is_void_v<R2>), "resolve/reject methods must return a promise of the same type or nothing");
         using LambdaReturnType = decltype(std::declval<LambdaReturnTrait>().template lambda<R1>());
 
-        return whenSettled(targetQueue, [resolveFunction = WTFMove(resolveFunction), rejectFunction = WTFMove(rejectFunction)] (ResultParam result) -> LambdaReturnType {
+        return whenSettled(targetQueue, [resolveFunction = std::forward<ResolveFunction>(resolveFunction), rejectFunction = std::forward<RejectFunction>(rejectFunction)] (ResultParam result) mutable -> LambdaReturnType {
             if (result) {
-                if constexpr (TakesArgument<ResolveFunction>::value)
-                    return resolveFunction(maybeMove(result.value()));
-                else
-                    return resolveFunction();
-            } else {
-                if constexpr (TakesArgument<RejectFunction>::value)
-                    return rejectFunction(maybeMove(result.error()));
+                if constexpr (std::is_void_v<ResolveValueT>)
+                    return invokeWithVoidOrWithArg(WTFMove(resolveFunction), nullptr_t());
```

Comment:
What is the point of calling invokeWithVoidOrWithArg vs. resolveFunction()?
It seems less readable.

Reply:
it handles automatically if the callback takes argument or not.
otherwise I would need that big if else statement we had before. it's much shorter that way.

plus it properly deals with moving the method or not which we didn't before